flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java

```diff
@@ -0,0 +1,156 @@
+			return r1.isPeriodic ? 1 : -1;
+		} else {
+			return Long.compare(r1.timestamp, r2.timestamp);
+		}
+	});
+
+	CheckpointRequestDecider(
+			int maxConcurrentCheckpointAttempts,
+			Consumer<Long> rescheduleTrigger,
+			Clock clock,
+			long minPauseBetweenCheckpoints,
+			Supplier<Integer> pendingCheckpointsSizeSupplier,
+			Object lock) {
+		this.maxConcurrentCheckpointAttempts = maxConcurrentCheckpointAttempts;
+		this.rescheduleTrigger = rescheduleTrigger;
+		this.clock = clock;
+		this.minPauseBetweenCheckpoints = minPauseBetweenCheckpoints;
+		this.pendingCheckpointsSizeSupplier = pendingCheckpointsSizeSupplier;
+		this.lock = lock;
+	}
+
+	Optional<CheckpointTriggerRequest> chooseRequestToExecute(CheckpointTriggerRequest newRequest, boolean isTriggering, long lastCompletionMs) {
+		return chooseRequestToExecute(false, () -> newRequest, isTriggering, lastCompletionMs);
+	}
+
+	Optional<CheckpointTriggerRequest> chooseQueuedRequestToExecute(boolean isTriggering, long lastCompletionMs) {
+		return chooseRequestToExecute(true, triggerRequestQueue::peek, isTriggering, lastCompletionMs);
+	}
+
+	/**
+	 * Choose the next {@link CheckpointTriggerRequest request} to execute based on the provided candidate and the
+	 * current state. Acquires a lock and may update the state.
+	 * @return request to execute, if any.
+	 */
+	private Optional<CheckpointTriggerRequest> chooseRequestToExecute(
+			boolean isEnqueued,
+			Supplier<CheckpointTriggerRequest> candidateSupplier,
+			boolean isTriggering,
+			long lastCompletionMs) {
+		synchronized (lock) {
+			final CheckpointTriggerRequest candidate = candidateSupplier.get();
+			if (candidate == null) {
+				return Optional.empty();
+			} else if (isTriggering) {
+				return onCantTriggerNow(candidate, isEnqueued);
+			} else if (pendingCheckpointsSizeSupplier.get() >= maxConcurrentCheckpointAttempts) {
+				return onCantCheckpointNow(candidate, TOO_MANY_CONCURRENT_CHECKPOINTS, () -> { /* not used */ }, isEnqueued);
+			} else if (nextTriggerDelayMillis(lastCompletionMs) > 0) {
+				return onCantCheckpointNow(candidate, MINIMUM_TIME_BETWEEN_CHECKPOINTS, () -> rescheduleTrigger.accept(nextTriggerDelayMillis(lastCompletionMs)), isEnqueued);
+			} else if (candidate != triggerRequestQueue.peek()) {
+				triggerRequestQueue.offer(candidate);
+				return Optional.of(triggerRequestQueue.poll());
+			} else {
+				return Optional.of(isEnqueued ? triggerRequestQueue.poll() : candidate);
+			}
+		}
+	}
+
+	private Optional<CheckpointTriggerRequest> onCantTriggerNow(CheckpointTriggerRequest candidate, boolean isEnqueued) {
+		offerIfNeeded(candidate, isEnqueued);
+		return Optional.empty();
+	}
+
+	private Optional<CheckpointTriggerRequest> onCantCheckpointNow(
+			CheckpointTriggerRequest candidate,
+			CheckpointFailureReason dropReason,
+			Runnable postDrop,
+			boolean isEnqueued) {
+		if (candidate.props.forceCheckpoint()) {
+			return Optional.of(isEnqueued ? triggerRequestQueue.poll() : candidate);
```

Comment:
it's a bit confusing the invartiant, that `candidate` can be still on the queue or not yet. I'm not sure, but It might cleaned up the code a bit, if you polled the candidate from the the queue immediately. This could unify some code paths with not enqueued candidate.

Especially that this `isEnqueued ? triggerRequestQueue.poll() : candidate` condition is already duplicated and can be easily forgotten in case of future changes.

Reply:
Yes, we could eliminate `isEnqueued` logic by either always adding (new) requests to the queue; or by always polling (enqueued) requests from it.
But it would have some performance penalty (`log(n)`). As it depends on the configuration I opted for a more complex code here.

Reply:
Changed code to always go through the queue.